### ISSUE TITLE
EqualsWrongThing not raised when returning from if-else

### ISSUE THREAD

**Issue by csrma** [google/error-prone#5157]
The following code mistakenly matches variable `this.a` with `that.b` and `this.b` with `that.a` in the `equals()` method. Ideally, this should raise `EqualsWrongThing` but it is not doing so.

Below is a ready-to-scan file where `EqualsWrongThing` should be raised on line 19 and 21. However, if line 18-19 is removed/commented out, then `EqualsWrongThing` is correctly flagged on line 21.
```java
import java.util.Objects;

class Frobnicator {
    private int a;
    private int b;
    private String checker;

    public Frobnicator(int a, int b) {
        this.a = a;
        this.b = b;
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof Frobnicator that))
            return false;

        if (this.checker == null && that.checker == null)              // line - 18
            return a == that.b && b == that.a;                         // line - 19
        
        return a == that.a && b == that.a;                             // line - 21
    }

    @Override
    public int hashCode() {
        return Objects.hash(a, b);
    }

    public boolean showBug(Frobnicator other) {
        return this.equals(other);
    }
}

class EqualsWrongThingExample {
    public static void main(String[] args) {
        Frobnicator f1 = new Frobnicator(1, 2);
        Frobnicator f2 = new Frobnicator(1, 2);
        Frobnicator f3 = new Frobnicator(2, 1);

        System.out.println(f1.showBug(f2)); // Expected true, but due to bug, it returns false
        System.out.println(f1.showBug(f3)); // Expected false, but due to bug, it returns true
    }
}
```
**Expected Outcome:** Flags `EqualsWrongThing` on line 19 and 21.

ErrorProne: `2.40.0`
Dataflow-ErrorProne: `3.42.0`
Java: `JDK 21`
Run Method: `CLI`

**Comment by graememorgan** [comment]
That's hitting this branch, I think, which is expected to avoid a family of false positives: https://github.com/google/error-prone/blob/ef5073b3e2bdd09e27f2196d0db5508ff8967b7d/core/src/main/java/com/google/errorprone/bugpatterns/EqualsWrongThing.java#L120

Did you see this crop up in reality, or is this a contrived example? I think I'd struggle to be sure that equals method isn't correct as a human.

**Comment by csrma** [comment]
The example shared here is generated by a testing tool. However, I think it is not uncommon to have multiple return branches based on availability of certain fields in complex real-world scenarios.

Also, I'm aware the equals method has other limitations as well (does not match the `checker` field in either case). But I wanted to keep the example minimal with only the parts necessary to manifest the issue.

**Comment by graememorgan** [comment]
I don't think the issue is the branching, it's that comparing `(a, b) == (b, a)` is specifically allowed for order-independent equals methods, which we're specifically allowing.
